### ISSUE TITLE
Sun-Synch orbits not working as expected

### ISSUE THREAD

**Issue by superx458** [poliastro/poliastro#1139]
Hi, 

I ran into a couple issues with using the helio-synchronous method. Since I'm new to orbit modeling, it may be my fault, so happy to hear any clarifications! 

1. When using the helio-synchronous orbit method, the results give invalid solutions or solutions that don't make physical sense (instead or say, throwing an error and saying no such orbit exists that fulfills the formulas). For instance, in creating the orbit below, the orbit's radial velocity at apogee is non-zero, which doesn't make sense, as only the tangential component should contribute (which is the case if you test the regular elliptical orbit method you guys provide). Further, the radial velocity in this case seems to be 0 at 105 degrees, which again doesn't make sense, since it's not at apogee or perigee here. 

2. Further, the method requires you to input a, eccentricity, and inclination to create the orbit, but wouldn't it make sense to only input two of the parameters so that the third condition is left parametrized by the other to to be determined so you may find a valid solution? Otherwise it throws an error? Because the way it is currently set up, it doesn't tell you whether these three numbers "actually" create a sun-synched orbit or not. 

3. Finally, the orbit seems to go into the Earth at perigee; why doesn't it throw an error in this case, saying if the orbit perigee is less than the Earth's radius, it's not possible? 
4. RAAN is not a number, and I'm not sure why. It's a weird string of letters and numbers. 

Here's the code that gives a funky v_rad = 0 at non-apogee, nu != 180 or 0. 

a_heli = 16000 * u.km
ecc_heli = 0.6* u.one
i_heli = 0 * u.deg
op_heli = OrbitPlotter2D()

orb_heli_test = Orbit.heliosynchronous(Earth, a=a_heli, ecc= ecc_heli, inc=i_heli, argp=0 * u.deg, nu= -105.5585* u.deg)
print('period', orb_heli_test.period.to(u.hour))
print('velocity', orb_heli_test.v)
print('eccen', orb_heli_test.ecc)
print('true anomaly',orb_heli_test.nu)
print('semi major', orb_heli_test.a)
print('raan', orb_heli_test.raan)
print('argp', orb_heli_test.argp)
Here's a plot: 
![funky velocities](https://user-images.githubusercontent.com/66441215/112727733-48d7c300-8efa-11eb-96d5-cb9a4f26184a.png)

Or maybe I misunderstand what's happening here. Again, I'm new to orbit modeling but happy to hear any misconceptions or clarifications!

**Comment by xlxs4** [comment]
Hey @superx458! Regarding RAAN being a weird string, does `orb_heli_test.raan.to(u.deg).value` cover you?

**Comment by xlxs4** [comment]
I am sorry, but I can't quite understand what you mean in point `2`. With that in mind:
* Have you checked the [documentation on `Orbit.heliosynchronous`](https://docs.poliastro.space/en/latest/autoapi/poliastro/twobody/orbit/index.html#poliastro.twobody.orbit.Orbit.heliosynchronous)?
  * The equations here I believe are derived by Vallado [1]:
     ![image](https://user-images.githubusercontent.com/54778816/112730554-ece15e80-8f3a-11eb-92c5-929c6449510c.png)
  * Also see this [GitHub issue](https://github.com/poliastro/poliastro/issues/514#issuecomment-517614759)
  * If you check the code, only two of the three parameters are required:
    ```python3
                if (a is None) and (ecc is None) and (inc is None):
                    # We check sufficient number of parameters
                    raise ValueError(
                        "At least two parameters of the set {a, ecc, inc} are required."
                    )
                elif a is None and (ecc is not None) and (inc is not None):
                    # Semi-major axis is the unknown variable
                    a = (
                        -3
                        * R_SSO ** 2
                        * J2_SSO
                        * np.sqrt(k_SSO)
                        / (2 * n_sunsync * (1 - ecc ** 2) ** 2)
                        * np.cos(inc)
                    ) ** (2 / 7)
                elif ecc is None and (a is not None) and (inc is not None):
                    # Eccentricity is the unknown variable
                    _ecc_0 = np.sqrt(
                        -3
                        * R_SSO ** 2
                        * J2_SSO
                        * np.sqrt(k_SSO)
                        * np.cos(inc.to(u.rad))
                        / (2 * a ** (7 / 2) * n_sunsync)
                    )
                    ecc = np.sqrt(1 - _ecc_0)
                elif inc is None and (ecc is not None) and (a is not None):
                    # Inclination is the unknown variable
                    inc = np.arccos(
                        -2
                        * a ** (7 / 2)
                        * n_sunsync
                        * (1 - ecc ** 2) ** 2
                        / (3 * R_SSO ** 2 * J2_SSO * np.sqrt(k_SSO))
                    )
    ```
  * if you mean that the current implementation doesn't verify that the input is feasible in a practical scenario, ignore the above

[1]: [Vallado, D. A. (2001). Fundamentals of astrodynamics and applications (Vol. 12). Springer Science & Business Media.](https://docs.poliastro.space/en/latest/bibliography.html#books-and-papers)

**Comment by superx458** [comment]
Hi @xlxs4 ! Thanks for the quick reply!! 

The RAAN thing you suggested did work- thanks (although unsure why it gives a value above 360 degrees). 

As for the 2 of 3 parameters, I cannot supply 2/3 (as in, after trying about 15 or so combinations of semi-major axis and inclination, it kept giving me "No SSO orbit with given parameters can be found."  - but when I give it three variables semi_major, inclination, and eccentricity, it happily gives me a plot  - why is that the case? And should I just hack away at more numbers for semi_major and inclination in the hopes of finding an SSO orbit with given parameters ?

**Comment by xlxs4** [comment]
> but when I give it three variables semi_major, inclination, and eccentricity, it happily gives me a plot 

from the `Orbit.heliosynchronous()` code:
```python3
        try:
            with np.errstate(invalid="raise"):
                if no parameter was given:
                     complain
                elif ecc and inc are given:
                    calculate a
                elif a and inc are given:
                    calculate ecc
                elif a and ecc are given:
                    calculate inc
        except FloatingPointError:
            raise ValueError("No SSO orbit with given parameters can be found.")
```
If I understand this correctly, if you give it all three parameters, there's no way the `ValueError` can be raised

**Comment by superx458** [comment]
![image](https://user-images.githubusercontent.com/66441215/112733158-a416af00-8f14-11eb-814f-7acf422e8b8d.png)

This image was particularly useful, from the GitHub article you mentioned. I now see why it's giving me "No SSO", because one doesn't exist (if you can confirm this is the case that would be educationally useful haha). 

Ah, interesting. Might be something to just add in the next iteration to prevent dumb ppl like me from being confused over why it's giving funky numbers. I wonder why it's giving some solution then.

**Comment by xlxs4** [comment]
It's great that the figure helped! It's also from Vallado, btw. Here's the caption, too:
![image](https://user-images.githubusercontent.com/54778816/112734652-c4b02a80-8f4f-11eb-9d7b-ad98ec2533eb.png)
I think a input validation could happen for this as well :)
Also, I'd like it if the docstrings containing the 3 formulae also referenced Vallado as a source, to avoid people getting confused and saving whoever's interested some time searching around.

> As for the 3 inputs thing: Ah, interesting. Might be something to just add in the next iteration to prevent dumb ppl like me from being confused over why it's giving funky numbers. I wonder why it's giving some solution then.

Do you have any suggestion in mind @superx458?

>  I wonder why it's giving some solution then.

If all 3 inputs are provided, all `heliosynchronous()` does is:
```python3
ss = cls.from_classical(
attractor, a, ecc, inc, raan, argp, nu, epoch=epoch.tdb, plane=plane
)
return ss
```
Where `from_classical()` does this basic input verification:
```python3
for element in a, ecc, inc, raan, argp, nu, epoch:
    if not element.isscalar:
        raise ValueError(f"Elements must be scalar, got {element}")

if ecc == 1.0 * u.one:
    raise ValueError("For parabolic orbits use Orbit.parabolic instead")

if not 0 * u.deg <= inc <= 180 * u.deg:
    raise ValueError("Inclination must be between 0 and 180 degrees")

if ecc > 1 and a > 0:
    raise ValueError("Hyperbolic orbits have negative semimajor axis")

if not -np.pi * u.rad <= nu < np.pi * u.rad:
    warn("Wrapping true anomaly to -π <= nu < π", stacklevel=2)
    nu = ((nu + np.pi * u.rad) % (2 * np.pi * u.rad) - np.pi * u.rad).to(
        nu.unit
    )
```

... which then, in turn, returns:
```python3
ss = ClassicalState(
    attractor, a * (1 - ecc ** 2), ecc, inc, raan, argp, nu, plane
)
return cls(ss, epoch)
```
where the `ClassicalState` also doesn't include any kind of testing. 
Long story short: nothing prevents poliastro from baking an orbit for you in this particular scenario. Whether this should be the case is another topic.

> I now see why it's giving me "No SSO", because one doesn't exist in the cases of loops I was doing (if you can confirm this is the case that would be educationally useful haha).

Hahaha, sure! Just need to know what the values were 😄

**Comment by xlxs4** [comment]
(@astrojuanlu: I've plotted the [test example from `test_orbit`]() and this is the resulting plot:
![image](https://user-images.githubusercontent.com/54778816/112735430-8a955780-8f54-11eb-81da-2fd4f7a9eae2.png)

is this related to https://github.com/poliastro/poliastro/issues/278?)

**Comment by xlxs4** [comment]
@superx458 sorry, forgot to reply to:
> The RAAN thing you suggested did work- thanks (although unsure why it gives a value above 360 degrees).

You can try [`wrap_at()`](https://docs.astropy.org/en/stable/api/astropy.coordinates.Angle.html#astropy.coordinates.Angle.wrap_at), as per the [`Angle` documentation](https://docs.astropy.org/en/stable/coordinates/angles.html#wrapping-and-bounds):

![image](https://user-images.githubusercontent.com/54778816/112735532-5c644780-8f55-11eb-88f1-6ad681d5e0b2.png)

**Comment by jorgepiloto** [comment]
Thanks for your support @xlxs4 🚀

I see the original code posted by @superx458 provides an inclination of zero degrees, which is a not valid parameter for a heliosynchronous oribt and thus, the code properly raised an exception. We could raise some additional information suggesting the reader to check SSO docs, I guess... 🤔

**Comment by xlxs4** [comment]
Do you think there should be some added input checks?

**Comment by superx458** [comment]
Good to know about the 3 inputs! I'll leave the added input checks stuff up to yall, yall are the code geniuses here; wrap_at worked as well!

I guess I need to go back to the drawing board and learn some more astrodynamics then, sorry! my PI told me a sun-synched orbit in the equatorial plane (i.e i=0) was possible, hence why I put it in; does the Vallado diagram I sent earlier imply that only solutions of 90 to 130 degrees are valid solutions, i guess? 

However, I still don't understand why in heliosynchronous(), valid solutions like a = 14000 and e = 0.6 give a nonzero value for velocity radial at apogee (nu = -pi) and instead gives a velocity radial = 0 at nu=-105 degrees (which doesn't make physical sense?)

**Comment by jorgepiloto** [comment]
Is it possible that you were actually interested in computing a `geosynchronous` orbit instead than a `heliosynchronous` one, @superx458? These kind of orbits are feasible for null inclination values.

Regarding the radial velocity component, it must be zero at both perigee and apogee. We do not provide any auxiliary function or method to compute those, how are you computing them? If you make use of flight path angle, it is shown that radial velocity at perigee is zero:

```python
import numpy as np
from astropy import units as u

from poliastro.bodies import Earth
from poliastro.twobody import Orbit

# Parameter definition
a_heli = 14000 * u.km
ecc_heli = 0.6 * u.one

# Build the orbit and propagate to periapsis (nu = 0 [deg])
sso = Orbit.heliosynchronous(Earth, a=a_heli, ecc=ecc_heli)
sso_at_perigee = sso.propagate_to_anomaly(0 * u.deg)

# Compute the position and velocity vectors and their norms
r_vec, v_vec = sso_at_perigee.rv()
r_norm, v_norm = [np.linalg.norm(vec) for vec in [r_vec, v_vec]]

# Solve the flight path angle
gamma = np.arcsin(np.dot(r_vec, v_vec) / (r_norm * v_norm))

# Retrieve radial and tangential components
v_n = v_vec * np.cos(gamma)
v_r = v_vec * np.sin(gamma)

# Print velocity components
print(f"Normal velocity component: {v_n}")
print(f"Radial velocity component: {v_r}")
```

which has the following output:

```
>>> Normal velocity component: [-6.38556238  2.26263762  8.24565975] km / s
>>> Radial velocity component: [-0.  0.  0.] km / s
```

**Comment by superx458** [comment]
I may have been interpreting the documentation on PoliAstro regarding what orbit.v returns. Your code shown made me realize that, so my bad. Thank you for all the help!

**Comment by astrojuanlu** [comment]
Hi all! Reading this after a few days, thanks @superx458 for the report and @xlxs4 and @jorgepiloto for the replies!

@superx458 To summarize: what would you suggest we could do, on the code or the documentation, to improve poliastro in this case?

**Comment by superx458** [comment]
> Hi all! Reading this after a few days, thanks @superx458 for the report and @xlxs4 and @jorgepiloto for the replies!
> 
> @superx458 To summarize: what would you suggest we could do, on the code or the documentation, to improve poliastro in this case?

Yeah , thank you guys for all the help! My PI is very happy with the results of my sims after you guys helped me debug the initial code, so thank you! 

If I can offer my two cents on what could be improved (just coming from someone who is using the package for the first time and doesn't know much about orbits):
1. Throwing an error when you input three parameters (a,inc,ecc), such that new users don't interpret the answer as a solution
2. Instituting a warning if an orbit solution is less than the radius of the orbiting object (i.e. if radius of perigee < r_earth it cant be a valid solution, which you have to manually filter out)
3. Clarifying what the velocity and radius vectors are in the documentation (that in order to get the magnitude of the velocity tangential or radial, you'd have to do some extra math). 

Those three would be helpful. Again, thank you!

**Comment by astrojuanlu** [comment]
These are all great suggestions! Will leave the issue open so we can implement them in the near future 🙌🏼
